File: 0110-balanced-binary-tree/0110-balanced-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
    if (!root) return true

    const leftNodeMaxDepth = getMaxDepth(root.left)
    const rightNodeMaxDepth = getMaxDepth(root.right)

    if ((Math.abs(leftNodeMaxDepth - rightNodeMaxDepth)) > 1) return false

    return isBalanced(root.left) && isBalanced(root.right)
};

function getMaxDepth(root: TreeNode | null): number {
    return !root ? 0 : (Math.max(getMaxDepth(root.left), getMaxDepth(root.right))) + 1
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;

        int leftNodeMaxDepth = getMaxDepth(root.left);
        int rightNodeMaxDepth = getMaxDepth(root.right);

        if (Math.abs(leftNodeMaxDepth - rightNodeMaxDepth) > 1) return false;

        return isBalanced(root.left) && isBalanced(root.right);
    }

    private int getMaxDepth(TreeNode node) {
        if (node == null) return 0;

        if (node.left == null && node.right == null) return 1;

        return Math.max(getMaxDepth(node.left), getMaxDepth(node.right)) + 1;
    }
}